src/main/java/com/endava/internship/service/UserServiceImpl.java

```diff
@@ -4,66 +4,84 @@
 import com.endava.internship.domain.User;
 import com.endava.internship.service.UserService;
 
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
+import java.util.*;
 import java.util.function.Function;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
 
 public class UserServiceImpl implements UserService {
 
     @Override
     public List<String> getFirstNamesReverseSorted(List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().map(User::getFirstName).
+                sorted(Comparator.reverseOrder()).collect(Collectors.toList());
     }
 
     @Override
     public List<User> sortByAgeDescAndNameAsc(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().sorted(Comparator.comparing(User::getAge).reversed().
+                        thenComparing(User::getFirstName)).
+                collect(Collectors.toList());
     }
 
     @Override
     public List<Privilege> getAllDistinctPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().flatMap(e -> e.getPrivileges().
+                stream()).collect(Collectors.toList());
```

Comment:
you should have only distinct Privileges. Add another user to test and it will fail

Reply:
this method was fixed